crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -847,6 +857,20 @@ impl<'db> TypeInferenceBuilder<'db> {
         self.infer_parameters(&function.parameters);
     }
 
+    fn infer_type_alias_type_params(&mut self, type_alias: &ast::StmtTypeAlias) {
+        let type_params = type_alias
+            .type_params
+            .as_ref()
+            .expect("type alias type params scope without type params");
+
+        self.infer_type_parameters(type_params);
+    }
+
+    fn infer_type_alias(&mut self, type_alias: &ast::StmtTypeAlias) {
+        // TODO: what should the deferred state be here?
+        self.infer_annotation_expression(&type_alias.value, DeferredExpressionState::Deferred);
```

Comment:
It should be `Deferred`, because the RHS of a type statement is lazily evaluated at runtime, meaning we should use end-of-scope name resolution rather than flow-sensitive name resolution for names in the expression.